### PR TITLE
[WIP] Compact memory proofs

### PR DESCRIPTION
**Pull request by vyomshm**
implements #17

### CODE REVIEW

**Comment by johannbarbie** [comment]
@vyomshm sry for the long silence. i've been out to a mediation retreat. can you help me understand how this code help to make compact proofs?

example:
- we have memory with 4 x 32 byte words
- the next opcode is MLOAD from word 1.
- now i want to initialize the memory with only the necessary word (1) and verify it is correct.

how can i do so?

**Comment by johannbarbie** [comment]
@vyomshm i see you marked the pr as [WIP]. is it still WIP?

**Comment by vyomshm** [comment]
@johannbarbie Just pushed my implementation.

I've restricted the memory size to multiples of two, and therefore one would need to provide atleast 2 words. 

After the operation has been executed using `EthereumRuntime`, the `Result` output's memory will be merklized and a `rootHash` generated. In combination with a `bytes32[] merkleProof` and the actual `rootHash`, the operation can be verified using openZeppelin's standard `MerkleProof` library.

File: contracts/mocks/CompactMemProofMock.sol [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    ) public view returns(bool){
+        uint[] memory accounts;
+        bytes memory accountsCode;
+        bytes32 logHash;
+        EthereumRuntime.Result memory res = ethRuntime.execute(
+            code, 
+            data, 
+            intInput, 
+            stack, 
+            mem,
```
Comment (johannbarbie) [inline contracts/mocks/CompactMemProofMock.sol]: I am confused about passing `mem` here.
I expected that we would only pass the `memProof`, feed it into the runtime, and if the runtime would modify the leaf, the proof/root would update. if the runtime would try to access any other element than the leaf in the memory, then the execution would throw/revert. 
The advantage would be that even if the memory is huge, we would only need to pass logarithmic size of data to the execution.
will try to prepare a proposal how to adjust the code.

Comment (vyomshm) [reply contracts/mocks/CompactMemProofMock.sol]: `mem` is the leaf value here. 

For example, say if you had a total memory size of 20 words but the operation you needed to verify only requires the 3rd word from the begining. 
In this case you would only pass in the first 4 words in the `mem`. The operation would be executed and a merkle Hash of the resulting memory calculated. This hash can then be verified using the provided `memProof`. If the resulting roothash equals `rootMemHash` then the party(solver or challenger)  has submitted the correct result.

**Comment by johannbarbie** [comment]
@vyomshm due to other merges i had to make these changes:
```
diff --git a/contracts/mocks/CompactMemProofMock.sol b/contracts/mocks/CompactMemProofMock.sol
index 32d68cf..fdbe230 100644
--- a/contracts/mocks/CompactMemProofMock.sol
+++ b/contracts/mocks/CompactMemProofMock.sol
@@ -1,6 +1,5 @@
-pragma experimental "v0.5.0";
+pragma solidity 0.5.2;
 pragma experimental ABIEncoderV2;
-pragma solidity 0.4.24;
 
 import {EthereumRuntime} from "../EthereumRuntime.sol";
 import {EVMMemory} from "../EVMMemory.slb";
@@ -16,12 +15,12 @@ contract CompactMemProofMock {
     }
 
     function verify(
-        bytes code,
+        bytes memory code,
         bytes memory data,
         uint[4] memory intInput,
         uint[] memory stack,
         bytes memory mem,
-        bytes32[] memProof,
+        bytes32[] memory memProof,
         bytes32 rootMemHash
     ) public view returns(bool){
         uint[] memory accounts;
diff --git a/contracts/mocks/MerkleProof.slb b/contracts/mocks/MerkleProof.slb
index cd095c6..a6b7ad1 100644
--- a/contracts/mocks/MerkleProof.slb
+++ b/contracts/mocks/MerkleProof.slb
@@ -1,4 +1,4 @@
-pragma solidity ^0.4.24;
+pragma solidity 0.5.2;
 
 /**
  * @title MerkleProof
@@ -13,7 +13,7 @@ library MerkleProof {
      * @param root Merkle root
      * @param leaf Leaf of Merkle tree
      */
-    function verify(bytes32[] proof, bytes32 root, bytes32 leaf) internal pure returns (bool) {
+    function verify(bytes32[] memory proof, bytes32 root, bytes32 leaf) internal pure returns (bool) {
         bytes32 computedHash = leaf;
 
         for (uint256 i = 0; i < proof.length; i++) {
diff --git a/test/TestEVMMemory.sol b/test/TestEVMMemory.sol
index 9575c46..93a94b4 100644
--- a/test/TestEVMMemory.sol
+++ b/test/TestEVMMemory.sol
@@ -194,6 +194,8 @@ contract TestEVMMemory {
     function testStoreBytesBigArray() public {
         bytes memory bts = new bytes(1513*32);
         EVMMemory.Memory memory mem = EVMMemory.newMemory();
+        uint fMem = MemOps.freeMemPtr();
+        uint tMem = mem.dataPtr + mem.cap * WORD_SIZE;
         mem.storeBytes(bts, 0, 0, bts.length);
         uint pos = mem.dataPtr;
         uint val;
```
but i can't apply then to your branch

File: contracts/mocks/CompactMemProofMock.sol (hunk 1 above)
Comment (johannbarbie) [reply contracts/mocks/CompactMemProofMock.sol]: if the code would require to read or write the 18 words, how big would `mem` need to be?

Comment (johannbarbie) [reply contracts/mocks/CompactMemProofMock.sol]: do you see a way to make this static sized `log(sizeIn32byteWords)` no matter which position is accessed? MSTORE and MLOAD opcodes only operate on a single 32 byte word. let's say we want to have memory size of multiple megabytes, then accessing the last element would exceed the gas limit to create a proof.

In this drawing i try to show that if you split the memory array into 32 byte slots, and merkelize them, writing to a proven leave anywhere in the tree only requires `log(sizeIn32byteWords)` elements.
https://docs.google.com/drawings/d/15NOju-EwJpn3JLX4zeAFvifIBGOY5D21BBwOc6cRRr0

**Comment by johannbarbie** [comment]
@vyomshm thanks for updating 👍

**Comment by vyomshm** [comment]
@johannbarbie Can you have a look ? I've updated the implementation accordingly + added all the relevant tests.

File: contracts/mocks/CompactMemProofMock.sol (hunk 1 above)
Comment (pinkiebell) [reply contracts/mocks/CompactMemProofMock.sol]: @vyomshm I the part above ^, involves more changes(a sparse memory setup), but I covered that in #58 .
The bounty scope was/is a bit chaotic, as the scope wasn't not clearly defined enough.
I'm going to implement your PR in a later state with the sparse memory setup and consider this issue done. Unless you have objections? 😄

Comment (vyomshm) [reply contracts/mocks/CompactMemProofMock.sol]: I see. That's okay with me 😄

Comment (johannbarbie) [reply contracts/mocks/CompactMemProofMock.sol]: @vyomshm submit on gitcoin :)

**Comment by johannbarbie** [comment]
closing i favor of @pinkiebell 's local checkout where he is trying to re-use some parts of it.

**Pull request closed without merging**